### ISSUE TITLE
NPE when serializing a `LocalDate` or `LocalDateTime` using `AsDeductionTypeSerializer`

### ISSUE THREAD

**Issue by mike-reynolds-savient** [FasterXML/jackson-modules-java8#296]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

In v2.14.2 AsDeductionTypeSerializer was introduced. 
Prior to this we could define an immutable interface as follows:

```java
    @Nullable
    @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION)
    /* Complex/ nested classes should be added to this list for Jackson to check */
    @JsonSubTypes({
            @JsonSubTypes.Type(StoreField.class),
            @JsonSubTypes.Type(StringArray.class),
            @JsonSubTypes.Type(NumberArray.class),
            @JsonSubTypes.Type(QueryRequest.class) })
    T getValue();
``` 

and set a value of LocalDate, LocalDateTime, OffsetDateTime etc as the return of `getValue()`

This would serialize correctly with `JsonMapper.builder().disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)` set.

We now get a "Cannot read field 'valueShape'" exception because `AsDeductionTypeSerializer`  returns null if the `WritableTypeId.valueShape` is not a structure start token.
This results in an NPE within the `LocalDateTimeSerializer` class on line 89;
```java
 if (typeIdDef.valueShape == JsonToken.START_ARRAY) {
``` 

### Version Information

2.14.2

### Reproduction

```java

@Value.Immutable
@JsonSerialize(as = ImmutableQueryPredicate.class)
@JsonDeserialize(as = ImmutableQueryPredicate.class)
public interface QueryPredicate<T> extends Serializable {

    @Nullable
    @JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION)
    /* Complex/ nested classes should be added to this list for Jackson to check */
    @JsonSubTypes({
            @JsonSubTypes.Type(StoreField.class),
            @JsonSubTypes.Type(StringArray.class),
            @JsonSubTypes.Type(NumberArray.class),
            @JsonSubTypes.Type(QueryRequest.class) })
    T getValue();

}


``` 
```java
    @Test
    void testSerialiseLocalDatePredicate() {

        final QueryPredicate<LocalDateTime> dateTimePredicate = ImmutableQueryPredicate.<LocalDateTime>builder()
                .value(LocalDateTime.of(2022, 06, 24, 10, 23, 51))
                .build();

        final String asString = mapper.writeValueAsString(dateTimePredicate);
}

``` 

### Expected behavior

If deduction cannot find the class it should return default property behaviour.

### Additional context

https://github.com/FasterXML/jackson-databind/issues/3711#issuecomment-1371580012

**Comment by cowtowncoder** [comment]
Ok, so, couple of things now that I re-read this again.

First: I can definitely avoid NPE for `LocalDate[Time]Serializer`, and that probably makes sense.

But second, I think that `DEDUCTION` based type deserializer can only work JSON Object represenations, not String-valued ones. So it can work to figure out type of POJOs, but not `LocalDate` or `LocalDateTime`. So ultimately I am not sure this can be made to work, if I understand use case correctly (I might misunderstand it tho).

**Comment by cowtowncoder** [comment]
Was able to reproduce (must force serialization as String), partial fix via #298, will try to cover all types for which NPE would be produced.

**Comment by cowtowncoder** [comment]
@mike-reynolds-savient I fixed the NPE part for 2.16(.2) -- although it will be a month or two until that release. If you can try 2.16.2-SNAPSHOT to see how things work that'd be good. I am not confident DEDUCTION will work in general, but at least it should not fail with NPE no matter what.

**Comment by mike-reynolds-savient** [comment]
Many thanks @cowtowncoder 
When 2.16.2 or 2.17 become available for download I'll give it a go again. (I can't get to the -SNAPSHOT)
